rpc, rest: Improve block rpc error handling, check header before attempting to read block data.

mzumsande opened the pull request:
Fixes #20978

If a block was pruned, `getblock` already returns a specific error: "Block not available (pruned data)".
But if we haven't received the full block yet (e.g. in a race with block downloading after a new block was received headers-first, or during IBD) we just return an unspecific "Block not found on disk" error and log
`ERROR: ReadBlockFromDisk: OpenBlockFile failed for FlatFilePos(nFile=-1, nPos=0) `
which suggest something went wrong even though this is a completely normal and expected situation.

This PR improves the error message and stops calling `ReadRawBlockFromDisk()`, when we already know from the header that the block is not available on disk.
Similarly, it prevents all other rpcs from calling blockstorage read functions unless we expect the data to be there, so that `LogError()` will only be thrown when there is an actual file system problem.

I'm not  completely sure if the cause is important enough to change the wording of the rpc error, that some scripts may rely on.
If reviewers prefer it, an alternative solution would be to keep returning the current "Block not found on disk" error, but return it immediately instead of calling `ReadRawBlockFromDisk`, which would at least prevent the log error and also be an improvement in my opinion.

mzumsande committed:
rpc: Improve getblock / getblockstats error when only header is available.

This improves the error message of the getblock and getblockstats rpc and prevents calls to
ReadRawBlockFromDisk(), which are unnecessary if we know
from the header nStatus field that the block is not available.

TheCharlatan commented:
Concept ACK

fjahr commented on src/rpc/blockchain.cpp:
```diff
@@ -606,6 +606,8 @@ static std::vector<uint8_t> GetRawBlockChecked(BlockManager& blockman, const CBl
         LOCK(cs_main);
         if (blockman.IsBlockPruned(blockindex)) {
             throw JSONRPCError(RPC_MISC_ERROR, "Block not available (pruned data)");
+        } else if (!(blockindex.nStatus & BLOCK_HAVE_DATA)) {
+            throw JSONRPCError(RPC_MISC_ERROR, "Block not available (not fully downloaded)");
         }
```
Here and in `rest.cpp`: Structuring it like this should save us the `IsBlockPruned()` call in the normal case (untested). `IsBlockPruned()` checks the same condition plus something extra so this should not change behavior.

```suggestion
        if (!(blockindex.nStatus & BLOCK_HAVE_DATA)) {
            if (blockman.IsBlockPruned(blockindex) {
                throw JSONRPCError(RPC_MISC_ERROR, "Block not available (pruned data)");
            }
            throw JSONRPCError(RPC_MISC_ERROR, "Block not available (not fully downloaded)");
        }
```

mzumsande replied:
done. Performance doesn't seem that important, but I also think it's a bit clearer that way.

fjahr commented:
Concept ACK

maflcko reviewed (commented):
Thanks for picking this up.

Seems fine to fix it this way.

However, you forgot `GetBlockChecked` (and all the other functions that can fail in the same way)?

My preference to fix it would be to move the checks to the inside of the block manager. This way,

* module-external cs_main locking (and races) are avoided, because a prune call happening after the  `IsBlockPruned` check (and leaving the scope of cs_main) could be detected, if wanted, by doing the check again in the block manager
* error case enumeration and handling are done in a single place, as opposed to every call site
* It would be easier for call sites to differentiate between read failures caused by logic bugs (what this pull is trying to solve) and read failures caused by IO corruption.

However, this basically requires a full rewrite of the blockmanager read-interface, so I haven't completed it yet. (Happy to pick it up again, if you think it makes sense as well)

mzumsande commented:
> However, you forgot GetBlockChecked (and all the other functions that can fail in the same way)?

Updated to change all functions I could find. I was only looking at `getblock` before, not at other RPCs.

>My preference to fix it would be to move the checks to the inside of the block manager. 

> However, this basically requires a full rewrite of the blockmanager read-interface, so I haven't completed it yet. (Happy to pick it up again, if you think it makes sense as well)

Makes sense to me, so concept a c k to that! Depending on how large this PR is going to be and how long it's going to take to complete it, I could either close this simple PR, which changes return errors and is not a refactor, or it could be a pure refactor on top of it? Let me know what you prefer.

tdb3 commented on src/rpc/blockchain.cpp:
```diff
@@ -583,8 +583,11 @@ static CBlock GetBlockChecked(BlockManager& blockman, const CBlockIndex& blockin
+            if (blockman.IsBlockPruned(blockindex)) {
+                throw JSONRPCError(RPC_MISC_ERROR, "Block not available (pruned data)");
+            }
+            throw JSONRPCError(RPC_MISC_ERROR, "Block not available (not fully downloaded)");
```
nit:  The comment below should also be adjusted.  The case where the header is available but we don't have the block yet is covered by this new `JSONRPCError`.

        // Block not found on disk. This could be because we have the block
        // header in our index but not yet have the block or did not accept the
        // block. Or if the block was pruned right after we released the lock above.

mzumsande replied:
done, thanks!

tdb3 commented on src/rpc/blockchain.cpp:
```diff
@@ -604,8 +607,11 @@ static std::vector<uint8_t> GetRawBlockChecked(BlockManager& blockman, const CBl
+            if (blockman.IsBlockPruned(blockindex)) {
+                throw JSONRPCError(RPC_MISC_ERROR, "Block not available (pruned data)");
+            }
+            throw JSONRPCError(RPC_MISC_ERROR, "Block not available (not fully downloaded)");
```
nit: Same for here

        // Block not found on disk. This could be because we have the block
        // header in our index but not yet have the block or did not accept the
        // block. Or if the block was pruned right after we released the lock above.

mzumsande replied:
done, thanks!

tdb3 reviewed (commented):
Concept ACK
Thanks for picking this up.  Overall, I think adding clarity for the user is a net positive.

> I'm not completely sure if the cause is important enough to change the wording of the rpc error, that some scripts may rely on.

I agree, this may be a borderline case because `Block not found on disk` seems like it is correct albeit not as descriptive.  Since we're not changing RPC data structure or error codes (just the message), I'm not even sure if implementing `deprecatedrpc` is fully appropriate.

https://corecheck.dev/bitcoin/bitcoin/pulls/30410

Looks like we lost coverage for the original `Block not found on disk` error, so it would be good to add a commit with a test for that (or in a follow up).  Maybe something like the following (and possibly move `move_block_file()` into the framework to allow reuse).

```diff
diff --git a/test/functional/rpc_blockchain.py b/test/functional/rpc_blockchain.py
index 4807551c0c3..363255ba48f 100755
--- a/test/functional/rpc_blockchain.py
+++ b/test/functional/rpc_blockchain.py
@@ -638,6 +638,11 @@ class BlockchainTest(BitcoinTestFramework):
         node.submitheader(block["hex"])
         assert_raises_rpc_error(-1, "Block not available (not fully downloaded)", lambda: node.getblock(block["hash"]))
 
+        self.log.info("Test getblock when block is missing")
+        move_block_file('blk00000.dat', 'blk00000.dat.bak')
+        assert_raises_rpc_error(-1, "Block not found on disk", node.getblock, blockhash)
+        move_block_file('blk00000.dat.bak', 'blk00000.dat')
+
 
 if __name__ == '__main__':
     BlockchainTest().main()
diff --git a/test/functional/rpc_getblockstats.py b/test/functional/rpc_getblockstats.py
index bf261befcc8..c4b4de65e80 100755
--- a/test/functional/rpc_getblockstats.py
+++ b/test/functional/rpc_getblockstats.py
@@ -182,5 +182,10 @@ class GetblockstatsTest(BitcoinTestFramework):
         assert_equal(tip_stats["utxo_increase_actual"], 4)
         assert_equal(tip_stats["utxo_size_inc_actual"], 300)
 
+        self.log.info('Test when block is missing')
+        (self.nodes[0].blocks_path / 'blk00000.dat').rename(self.nodes[0].blocks_path / 'blk00000.dat.backup')
+        assert_raises_rpc_error(-1, 'Block not found on disk', self.nodes[0].getblockstats, hash_or_height=1)
+        (self.nodes[0].blocks_path / 'blk00000.dat.backup').rename(self.nodes[0].blocks_path / 'blk00000.dat')
+
 if __name__ == '__main__':
     GetblockstatsTest().main()
```

andrewtoth commented on src/rpc/blockchain.cpp:
```diff
@@ -629,8 +635,11 @@ static CBlockUndo GetUndoChecked(BlockManager& blockman, const CBlockIndex& bloc
-            throw JSONRPCError(RPC_MISC_ERROR, "Undo data not available (pruned data)");
+        if (!(blockindex.nStatus & BLOCK_HAVE_DATA)) {
+            if (blockman.IsBlockPruned(blockindex)) {
+                throw JSONRPCError(RPC_MISC_ERROR, "Block not available (pruned data)");
```
```suggestion
                throw JSONRPCError(RPC_MISC_ERROR, "Undo not available (pruned data)");
```

mzumsande replied:
fixed (see below)

andrewtoth commented on src/rpc/blockchain.cpp:
```diff
@@ -629,8 +635,11 @@ static CBlockUndo GetUndoChecked(BlockManager& blockman, const CBlockIndex& bloc
+            if (blockman.IsBlockPruned(blockindex)) {
+                throw JSONRPCError(RPC_MISC_ERROR, "Block not available (pruned data)");
+            }
+            throw JSONRPCError(RPC_MISC_ERROR, "Block not available (not fully downloaded)");
```
```suggestion
            throw JSONRPCError(RPC_MISC_ERROR, "Undo not available (not fully downloaded)");
```

mzumsande replied:
fixed (see below)

andrewtoth commented on src/rpc/blockchain.cpp:
```diff
@@ -629,8 +635,11 @@ static CBlockUndo GetUndoChecked(BlockManager& blockman, const CBlockIndex& bloc
+            if (blockman.IsBlockPruned(blockindex)) {
+                throw JSONRPCError(RPC_MISC_ERROR, "Block not available (pruned data)");
+            }
+            throw JSONRPCError(RPC_MISC_ERROR, "Block not available (not fully downloaded)");
```
It feels like the rule of the 3 is being hit here and this code could be DRY'd up? Maybe in a helper function that takes the string `"Block"` or `"Undo"` and does string formatting?

mzumsande replied:
I have now added a single function `CheckHeaderForData` that takes a bool for querying undo data.

andrewtoth reviewed (approved):
Concept ACK

andrewtoth commented on src/rpc/blockchain.cpp:
```diff
@@ -629,8 +635,11 @@ static CBlockUndo GetUndoChecked(BlockManager& blockman, const CBlockIndex& bloc
         LOCK(cs_main);
-        if (blockman.IsBlockPruned(blockindex)) {
-            throw JSONRPCError(RPC_MISC_ERROR, "Undo data not available (pruned data)");
+        if (!(blockindex.nStatus & BLOCK_HAVE_DATA)) {
```
```suggestion
        if (!(blockindex.nStatus & BLOCK_HAVE_UNDO)) {
```

mzumsande replied:
fixed

mzumsande committed:
rest: improve error when only header of a block is available.

This avoids calling ReadRawBlockFromDisk() when the block is expected
not to be available because we haven't downloaded it yet and only know
the header.

mzumsande committed:
rpc: check block index before reading block / undo data

This avoids low-level log errors that are supposed to only occur when
there is an actual problem with the block on disk missing unexpectedly,
but not in the case where the block and/or undo data are expected not to be there.

It changes behavior such that in the first case (block index indicates
data is available but retrieving it fails) an error is thrown.

It also adjusts a functional tests that tried to simulate not
having undo data (but having block data) by deleting the undo file.
This situation should occur reality because block and undo data are pruned together.
Instead, test this situation with a block that hasn't been connected.

mzumsande commented:
> Ok, I am done, but I think it is easier to review this pull request.

OK - I'm working on addressing the feedback, plus adding more test coverage for the different types of errors. Will update soon!

mzumsande committed:
test: add coverage to getblock and getblockstats

also removes an unnecessary newline.

Co-authored-by: tdb3 <106488469+tdb3@users.noreply.github.com>

BrandonOdiwuor reviewed (commented):
Concept ACK

mzumsande committed:
rpc: Improve gettxoutproof error when only header is available.

andrewtoth reviewed (approved):
utACK 02f6a42be3dfc57998304801c0d909bfa96745e4

mzumsande commented:
[02f6a42 ](https://github.com/bitcoin/bitcoin/commit/02f6a42be3dfc57998304801c0d909bfa96745e4)to 4fb708c:
Rebased and fixed silent conflict with #30636.

andrewtoth commented:
I'm unsure why CI is failing now...

fanquake commented:
@andrewtoth the CI failure in the test-each-commit job is unrelated to the changes (missing zmq package, fixed in #30749).

mzumsande commented:
> @andrewtoth the CI failure in the test-each-commit job is unrelated to the changes (missing zmq package, fixed in #30749).

Thanks, I also wasn't sure! I've just rebased again.

fjahr commented on src/rpc/blockchain.cpp:
```diff
@@ -184,7 +184,7 @@ UniValue blockToJSON(BlockManager& blockman, const CBlock& block, const CBlockIn
             CBlockUndo blockUndo;
             const bool is_not_pruned{WITH_LOCK(::cs_main, return !blockman.IsBlockPruned(blockindex))};
-            const bool have_undo{is_not_pruned && blockman.UndoReadFromDisk(blockUndo, blockindex)};
+            const bool have_undo{is_not_pruned && WITH_LOCK(::cs_main, return blockindex.nStatus & BLOCK_HAVE_UNDO) && blockman.UndoReadFromDisk(blockUndo, blockindex)};
```
In 041b78ccede9974117dac5d8623fedec2f32471d:

nit: I think it would be cleaner and easier to read if you break this up.

```suggestion
            const bool have_undo{is_not_pruned && WITH_LOCK(::cs_main, return blockindex.nStatus & BLOCK_HAVE_UNDO)};
            if (have_undo) {
                blockman.UndoReadFromDisk(blockUndo, blockindex);
            }
```

(same in `rpc/rawtransaction.cpp`)

mzumsande replied:
I can break this up, but I think the suggestion doesn't work because we still need to feed the result into `have_undo` so that if `UndoReadFromDisk` fails for an unexpected reason, we don't attempt to access `blockUndo` below and segfault.

fjahr replied:
> I think the suggestion doesn't work because we still need to feed the result into have_undo so that if UndoReadFromDisk fails for an unexpected reason like disk errors, we don't attempt to access blockUndo below and segfault.

I see, I think in that scenario it would be best to throw an error right there instead of ignoring it silently and giving the user an incomplete response? I mean, we have all indication that we should have the undo data in that scenario and then we can not read it, it seems like an error that the user should know about.

> In rpc/rawtransaction.cpp, did you mean to introduce a variable like skip_details or something to break this up?

I think that would work, I don't remember if I had something explicit in mind when I wrote it. It's just an awfully long list of conditions already and summing parts of them up in a variable with a concise name would make it easier to parse. I have left two additional suggestions there for shortening it inline and I think there it's the same question if we really want to silently return a limited result if we fail to read from disk what we are supposed to have. I'll wait for you feedback on that part and my inline suggestions. If nothing else I think it would also help a lot to just add line breaks after every condition.

mzumsande replied:
I think it makes sense to throw an error in this situation. Although it's not necessarily due to disk corruption, it would also be possible that the block/undo data is pruning event right after the blockindex check (`cs_main` is released) and before the `UndoReadFromDisk` call.

We also have a functional test for just this situation that wouldn't work anymore ([here](https://github.com/bitcoin/bitcoin/blob/c66c68345efb0bb3d5613ebac703cde779fa0f01/test/functional/rpc_blockchain.py#L615-L628)) but that tests doesn't actually prune a block but deletes an undo file to simulate pruning. So we'd either have to just remove it or actually prune a block... I'll try this out in the next days.

mzumsande commented on src/rpc/blockchain.cpp:
```diff
@@ -580,7 +580,7 @@ static RPCHelpMan getblockheader()
     };
 }
 
-static void CheckHeaderForData(BlockManager& blockman, const CBlockIndex& blockindex, bool check_for_undo)
```
done

mzumsande commented on src/rpc/blockchain.cpp:
```diff
@@ -580,20 +580,32 @@ static RPCHelpMan getblockheader()
     };
 }
 
+static void CheckHeaderForData(BlockManager& blockman, const CBlockIndex& blockindex, bool check_for_undo)
```
renamed to `CheckBlockDataAvailability()`

mzumsande commented:
[041b78c ](https://github.com/bitcoin/bitcoin/commit/041b78ccede9974117dac5d8623fedec2f32471d)to [92236f4](https://github.com/bitcoin/bitcoin/commit/92236f43ff92c931f3a099e03d7851b890bff263): addressed feedback by @fjahr

fjahr reviewed (commented):
@mzumsande Sorry, I had written the comment below already earlier but failed to submit it because of github being github. But it seems like it's in line with your response.

furszy commented on src/rpc/rawtransaction.cpp:
```diff
@@ -405,7 +405,7 @@ static RPCHelpMan getrawtransaction()
 
-    if (tx->IsCoinBase() || !blockindex || WITH_LOCK(::cs_main, return chainman.m_blockman.IsBlockPruned(*blockindex)) ||
+    if (tx->IsCoinBase() || !blockindex || WITH_LOCK(::cs_main, return !(blockindex->nStatus & BLOCK_HAVE_UNDO) || !(blockindex->nStatus & BLOCK_HAVE_DATA)) ||
         !(chainman.m_blockman.UndoReadFromDisk(blockUndo, *blockindex) && chainman.m_blockman.ReadBlockFromDisk(block, *blockindex))) {
```
q: if the `block` is only used to obtain the undo data, shouldn't this be a `!chainman.m_blockman.ReadBlockFromDisk(block, *blockindex))`? (negated)

mzumsande replied:
We need both block and undo data below if we don't skip with the non-detailed response here. So there is a parenthesis after the first negation, so the logic is `not(a and b)` which is the same as `not(a) or not(b).`

furszy replied:
ah, missed the extra parenthesis. Good.
Might be good to decouple it as proposed in https://github.com/bitcoin/bitcoin/pull/30410#discussion_r1752283015.

I wish there would be methods for "hasBlockData()" and "hasUndoData()" so we are not forced to use the internal `BlockStatus` enum everywhere.

furszy reviewed (commented):
Code review ACK 92236f43ff92c931f3a099e03d7851b890bff263

mzumsande commented on src/rpc/rawtransaction.cpp:
```diff
@@ -405,7 +405,7 @@ static RPCHelpMan getrawtransaction()
     CBlock block;
 
-    if (tx->IsCoinBase() || !blockindex || WITH_LOCK(::cs_main, return chainman.m_blockman.IsBlockPruned(*blockindex)) ||
+    if (tx->IsCoinBase() || !blockindex || WITH_LOCK(::cs_main, return !(blockindex->nStatus & BLOCK_HAVE_UNDO) || !(blockindex->nStatus & BLOCK_HAVE_DATA)) ||
```
done, good idea!

mzumsande commented:
[92236f4 ](https://github.com/bitcoin/bitcoin/commit/92236f43ff92c931f3a099e03d7851b890bff263)to [ccee8a2](https://github.com/bitcoin/bitcoin/commit/ccee8a274770a060fa917f24828f2e2f5c1875ee):

Changed the last commit such that now it throws an error if our block index indicates we should have block/undo data but we can't read it.
Also changed the functional test: Before, it would try to imitate a "pruning" situation by just deleting the undo file. This made little sense to me, because when we prune we prune both block and undo data together (and would have failed earlier due to missing block data). The situation with block data and no undo data can occur though, with blocks that haven't been connected. I've added a test for that.

fjahr commented on src/rpc/blockchain.cpp:
```diff
@@ -597,20 +597,32 @@ static RPCHelpMan getblockheader()
+void CheckBlockDataAvailability(BlockManager& blockman, const CBlockIndex& blockindex, bool check_for_undo)
+{
+    AssertLockHeld(cs_main);
+    uint32_t flag = check_for_undo ? BLOCK_HAVE_UNDO : BLOCK_HAVE_DATA;
```
More or a question: when `check_for_undo` is true doe we actually want to check for both? Having undo and not data set here would seem to be something going weirdly wrong though I guess.

```suggestion
    uint32_t flag = check_for_undo ? BLOCK_HAVE_MASK : BLOCK_HAVE_DATA;
```

mzumsande replied:
I agree that it shouldn't be possible to have undo data but no block data.
On the other hand, callers of this function may not even be interested in the block data, and I'm not sure if a RPC call is the right place to check for this inconsistency. At least it's checked in `CheckBlockIndex()` [here](https://github.com/bitcoin/bitcoin/blob/06a9f7789e359131abad2489a221888f711bdc3b/src/validation.cpp#L5326), so I think I'll leave this as is for now.

fjahr commented on src/rpc/rawtransaction.cpp:
```diff
@@ -405,11 +405,16 @@ static RPCHelpMan getrawtransaction()
         return result;
     }
+    if (!chainman.m_blockman.UndoReadFromDisk(blockUndo, *blockindex)) {
+        throw JSONRPCError(RPC_INTERNAL_ERROR, "Undo data expected but can't be read. This could be due to disk coruption or a conflict with a pruning event.");
```
typo: corruption

mzumsande replied:
fixed

fjahr commented on src/rpc/rawtransaction.cpp:
```diff
@@ -405,11 +405,16 @@ static RPCHelpMan getrawtransaction()
+        throw JSONRPCError(RPC_INTERNAL_ERROR, "Undo data expected but can't be read. This could be due to disk coruption or a conflict with a pruning event.");
+    }
+    if (!chainman.m_blockman.ReadBlockFromDisk(block, *blockindex)) {
+        throw JSONRPCError(RPC_INTERNAL_ERROR, "Block data expected but can't be read. This could be due to disk coruption or a conflict with a pruning event.");
```
typo: corruption

fjahr commented on src/rpc/blockchain.cpp:
```diff
@@ -201,8 +201,10 @@ UniValue blockToJSON(BlockManager& blockman, const CBlock& block, const CBlockIn
-
+            bool have_undo{is_not_pruned && WITH_LOCK(::cs_main, return blockindex.nStatus & BLOCK_HAVE_UNDO)};
+            if (have_undo && !blockman.UndoReadFromDisk(blockUndo, blockindex)) {
+                throw JSONRPCError(RPC_INTERNAL_ERROR, "Undo data expected but can't be read. This could be due to disk coruption or a conflict with a pruning event.");
```
typo: corruption

fjahr reviewed (commented):
Code review ACK a5df908d981138f066f5f963a1813d453483814a

Feel free to ignore typo comments unless you retouch.

Also saw a typo in 8bfb0bd7bd5f8963fb1f58988f9bb36199dc98de commit message: "dowloaded"

mzumsande commented:
[a5df908 ](https://github.com/bitcoin/bitcoin/commit/a5df908d981138f066f5f963a1813d453483814a)to [6a1aa51](https://github.com/bitcoin/bitcoin/commit/6a1aa510e31e8b77793341473aa5afc9d023a6e3): Fixed typos.

andrewtoth reviewed (approved):
re-ACK 6a1aa510e31e8b77793341473aa5afc9d023a6e3

fjahr commented:
re-ACK 6a1aa510e31e8b77793341473aa5afc9d023a6e3

achow101 commented:
ACK 6a1aa510e31e8b77793341473aa5afc9d023a6e3

Sjors commented:
It would be nice to have unique (new) RPC error codes for this, because e.g. with ForkMonitor I was parsing the string that comes with `RPC_MISC_ERROR`, which this PR slightly changes.

https://github.com/jonathanbier/forkmonitor/commit/2d0e9f48cbd95b5ce52416ef497d1691ea3e24f0
